x/committee/types/permissions.go

```diff
@@ -122,14 +123,149 @@ func (changes AllowedParamsChanges) FilterByParamChange(paramChange paramspropos
+		if req == nil {
+			return false
+		}
+
+		for _, v := range incomingRecords {
+			if v[req.Key] == req.Val {
+				incoming = &v
+				break
+			}
+		}
+
+		// disallow the change if no incoming record found for current record.
+		if incoming == nil {
+			return false
+		}
+
+		// check incoming changes are allowed
+		allowed := validateParamChangesAreAllowed(current, *incoming, req.AllowedSubparamAttrChanges)
+
+		if !allowed {
+			return false
+		}
+	}
+
+	return true
+}
+
+func validateParamChangesAreAllowed(current SubparamChanges, incoming SubparamChanges, allowList []string) bool {
+	// make sure we are not adding or removing any new attributes
+	if len(current) != len(incoming) {
+		return false
+	}
+
+	for k, v := range current {
+		isAllowed := false
+
+		// check if the param attr key is in the allow list
+		for _, allowedKey := range allowList {
+			if k == allowedKey {
+				isAllowed = true
+				break
+			}
+		}
+
+		// if not allowed, incoming value needs to be the same, or it is rejected
+		if !isAllowed {
+			// since we cannot compare maps directly, we need to convert them to json first.
+			// this should be fine since the data we are marshalling here should always be pretty small.
+			if reflect.TypeOf(v).Kind() == reflect.Map {
+				data, err := json.Marshal(v)
+				if err != nil {
+					return false
+				}
+				data2, err := json.Marshal(incoming[k])
+				if err != nil {
+					return false
+				}
+				if string(data) != string(data2) {
+					return false
+				}
+			} else if v != incoming[k] {
+				return false
+			}
+		}
+	}
+
+	return true
+}
+
+func (allowed AllowedParamsChange) allowsSingleParamsChange(current SubparamChanges, incoming SubparamChanges) bool {
+	return validateParamChangesAreAllowed(current, incoming, allowed.SingleSubparamAllowedAttrs)
+}
+
+// allowsParamChange returns true if the given proposal param change is allowed by the AllowedParamsChange rules.
+func (allowed AllowedParamsChange) allowsParamChange(ctx sdk.Context, paramsChange paramsproposal.ParamChange, pk ParamKeeper) bool {
 	// Check if param change matches target subspace and key.
 	if allowed.Subspace != paramsChange.Subspace && allowed.Key != paramsChange.Key {
 		return false
 	}
 
-	// TODO: Handle sub parameters and required sub param attr values
+	// Check if param value is an array before unmarshalling to corresponding types
+	tdata := strings.TrimLeft(paramsChange.Value, "\t\r\n")
+	isArray := len(tdata) > 0 && tdata[0] == '['
 
-	return true
+	// Handle multi param value validation
+	if isArray {
+		var changeValue MultiSubparamChanges
+		if err := json.Unmarshal([]byte(paramsChange.Value), &changeValue); err != nil {
+			return false
+		}
+
+		var currentValue MultiSubparamChanges
+		subspace, found := pk.GetSubspace(paramsChange.Subspace)
+		if !found {
+			return false
+		}
+		raw := subspace.GetRaw(ctx, []byte(paramsChange.Key))
+		if err := json.Unmarshal(raw, &currentValue); err != nil {
+			panic(err)
```

Comment:
panic here to match `subspace.Get` behaviour. We should never have bad data stored in subspace so I think panic here make sense.